jscomp/syntax/ast_tdcls.ml

```diff
@@ -44,6 +44,30 @@ let newTdcls
          else x )
       
 
+#if BS_NATIVE then
+
+let turn_bs_optional_into_optional (tdcls : Parsetree.type_declaration list) =
+   List.map (fun tdcl -> match tdcl.Parsetree.ptype_kind with
+   | Ptype_record labels ->
+     {tdcl with ptype_kind = Ptype_record (List.map (fun ({Parsetree.pld_type; pld_loc; pld_attributes} as dcl : Parsetree.label_declaration) ->
+           let has_optional_field = Ast_attributes.has_bs_optional pld_attributes in
+           if has_optional_field then
+             { dcl with
+               Parsetree.pld_type = {dcl.pld_type with ptyp_desc =
+                Ptyp_constr(
+                  {txt = Lident "option";
+                   loc = pld_loc}
+                   , [pld_type]);
+                   ptyp_loc = pld_loc;
+               };
+               pld_attributes = Ext_list.exclude pld_attributes (fun x -> (Ast_attributes.is_optional x) || (Ast_attributes.is_bs_as x))
+             }
+           else dcl
+         ) labels)}
+   | _ -> tdcl) tdcls
+
+#end
+
 
```

Comment:
This could be cleaned up a little bit, but I will merge it to save it for future work, thanks!